backend/server/src/scheduler/service.rs

```diff
@@ -0,0 +1,61 @@
+use super::builder::{BuildRequest, Builder};
+use super::ingress::{IngressService, IngressTask};
+use super::recorder::{RecorderService, RecorderTask};
+use crate::db::{model::drv, DbService};
+use tokio::sync::mpsc;
+use tokio::task::JoinHandle;
+
+/// SchedulerService spins up three smaller services:
+///   RequestIngress:
+///         Handles taking in many drv build requests and determines
+///         if the drv is "buildable", has a failed dependency, otherwise it's just queued
+///   Builder:
+///         Receives a stream of "buildable" drvs and builds them.
+///         Upon completion of a build, passes build status to RecorderService
+///   RecorderService:
+///         "Records" build status to the database.
+///         Upon recording a successful build, it pushes a check request for
+///         each downstream drv to RequestIngress service to see if the drv
+///         is now "buildable"
+#[allow(dead_code)]
+pub struct SchedulerService {
+    // Handle to the db service, useful for persisting and querying build state
+    db_service: DbService,
+
+    // Channels to individual services
+    // We may in the future need to recover an individual service, so retaining
+    // a handle to the other service channels will be prequisite
+    ingress_sender: mpsc::Sender<IngressTask>,
+    builder_sender: mpsc::Sender<BuildRequest>,
+    recorder_sender: mpsc::Sender<RecorderTask>,
+
+    ingress_thread: JoinHandle<()>,
+    builder_thread: JoinHandle<()>,
+    recorder_thread: JoinHandle<()>,
+}
+
+impl SchedulerService {
+    pub fn new(db_service: DbService) -> anyhow::Result<Self> {
+        let (ingress_service, ingress_sender) = IngressService::init(db_service.clone());
```

Comment:
did you try to avoid all these clones here? Seems a bit wasteful.

Reply:
it's very cheap since it's just a db handle, I thought the same thing as well.